File: luhn.py
#!/usr/bin/env python3
"""
luhn.py — Luhn Algorithm: Checksum Validation & Check Digit Generation

The Luhn (Mod 10) algorithm verifies the integrity of identification numbers
such as credit card numbers and IMEI identifiers.

Usage:
    python luhn.py validate   <number>          # validate a complete number
    python luhn.py checkdigit <partial_number>  # generate the check digit
    python luhn.py                              # interactive mode
"""

import sys
import argparse


# ---------------------------------------------------------------------------
# Card / number type identification
# ---------------------------------------------------------------------------

def identify_card_type(number: str) -> str:
    n, p2, p4, p6 = len(number), number[:2], number[:4], number[:6]

    if number[0] == "4" and n in (13, 16, 19):                          return "Visa"
    if number[0] == "5" and n == 16 and "51" <= p2 <= "55":             return "Mastercard"
    if number[0] == "2" and n == 16 and "222100" <= p6 <= "272099":     return "Mastercard (2-series)"
    if number[0] == "3" and n == 15 and p2 in ("34", "37"):             return "American Express"
    if number[0] == "3" and n == 14 and p2 in ("30", "36", "38"):       return "Diners Club"
    if n == 16 and (p4 == "6011" or "644" <= p4[:3] <= "649" or p2 == "65"): return "Discover"
    if n == 16 and p2 == "62":                                           return "UnionPay"
    if n in (15, 16):                                                    return "IMEI / Credit Card (unrecognized network)"
    return "Unknown"


# ---------------------------------------------------------------------------
# Core algorithm — single pass, prints walkthrough, returns result
# ---------------------------------------------------------------------------

def _run_luhn(digits: list[int], start_from_right_offset: int) -> tuple[list[int], int]:
    """
    Double every second digit starting at index (len-1 - start_from_right_offset)
    going left. Return the modified digit list and the total sum.
    """
    working = digits[:]
    i = len(working) - 1 - start_from_right_offset
    while i >= 0:
        working[i] *= 2
        if working[i] > 9:
            working[i] -= 9
        i -= 2
    return working, sum(working)


def validate_verbose(number: str) -> None:
    """Validate a complete number with step-by-step output."""
    digits = [int(c) for c in number]

    print(f"\nNumber type : {identify_card_type(number)}")
    print(f"Digit count : {len(number)}")
    print(f"Digits      : {' '.join(number)}")

    # Validation: double starting from second-to-last (offset = 1)
    step1 = digits[:]
    i = len(step1) - 2
    while i >= 0:
        step1[i] *= 2
        i -= 2
    print(f"\n[Step 1] Double every second digit from the right:\n         {' '.join(map(str, step1))}")

    step2 = [x - 9 if x > 9 else x for x in step1]
    print(f"\n[Step 2] Subtract 9 from values > 9:\n         {' '.join(map(str, step2))}")

    total = sum(step2)
    print(f"\n[Step 3] Total sum : {total}  |  mod 10 : {total % 10}")
    print(f"\nResult --> Luhn Checksum: {'VALID' if total % 10 == 0 else 'INVALID'}")


def checkdigit_verbose(partial: str) -> None:
    """Compute and display the check digit with step-by-step output."""
    print(f"\nPartial number : {' '.join(partial)}")
    print(f"Digit count    : {len(partial)}")

    # Check-digit generation: double starting from the rightmost digit (offset = 0)
    step1 = [int(c) for c in partial]
    i = len(step1) - 1
    while i >= 0:
        step1[i] *= 2
        i -= 2
    print(f"\n[Step 1] Double every second digit from the right:\n         {' '.join(map(str, step1))}")

    step2 = [x - 9 if x > 9 else x for x in step1]
    print(f"\n[Step 2] Subtract 9 from values > 9:\n         {' '.join(map(str, step2))}")

    total = sum(step2)
    check = (10 - (total % 10)) % 10
    print(f"\n[Step 3] Sum : {total}  |  Check digit = (10 - {total % 10}) mod 10 = {check}")
    print(f"\nResult --> Check Digit : {check}")
    print(f"           Full number : {partial}{check}")


# ---------------------------------------------------------------------------
# Public API (importable)
# ---------------------------------------------------------------------------

def luhn_validate(number: str) -> bool:
    """Return True if *number* (string of digits) passes the Luhn check."""
    if not number.isdigit():
        raise ValueError("Input must contain digits only.")
    digits = [int(c) for c in number]
    i = len(digits) - 2
    while i >= 0:
        digits[i] = digits[i] * 2 - (9 if digits[i] * 2 > 9 else 0)
        i -= 2
    return sum(digits) % 10 == 0


def luhn_check_digit(partial: str) -> int:
    """Return the check digit to append to *partial* to make it Luhn-valid."""
    if not partial.isdigit():
        raise ValueError("Input must contain digits only.")
    digits = [int(c) for c in partial]
    i = len(digits) - 1
    while i >= 0:
        digits[i] = digits[i] * 2 - (9 if digits[i] * 2 > 9 else 0)
        i -= 2
    return (10 - (sum(digits) % 10)) % 10


# ---------------------------------------------------------------------------
# CLI
# ---------------------------------------------------------------------------

def _interactive() -> None:
    print("=" * 48)
    print("  Luhn Algorithm — Interactive Mode")
    print("=" * 48)
    print("\n[1] Validate a complete number")
    print("[2] Generate a check digit")
    choice = input("\nChoice (1/2): ").strip()

    if choice == "1":
        raw = input("Enter the complete number: ").strip().replace(" ", "").replace("-", "")
        if not raw.isdigit() or not (13 <= len(raw) <= 19):
            print("Error: Must be 13–19 digits.")
            return
        validate_verbose(raw)

    elif choice == "2":
        raw = input("Enter the partial number (without check digit): ").strip().replace(" ", "").replace("-", "")
        if not raw.isdigit() or not (12 <= len(raw) <= 18):
            print("Error: Must be 12–18 digits.")
            return
        checkdigit_verbose(raw)

    else:
        print("Invalid choice.")


def main() -> None:
    parser = argparse.ArgumentParser(
        description="Luhn Algorithm — Checksum Validation & Check Digit Generation",
        epilog=(
            "Examples:\n"
            "  python luhn.py validate   4532015112830366\n"
            "  python luhn.py checkdigit 453201511283036\n"
            "  python luhn.py"
        ),
        formatter_class=argparse.RawDescriptionHelpFormatter,
    )
    sub = parser.add_subparsers(dest="cmd")

    p_val = sub.add_parser("validate", help="Validate a complete number")
    p_val.add_argument("number", help="Full number including check digit")

    p_cd = sub.add_parser("checkdigit", help="Generate a Luhn check digit")
    p_cd.add_argument("partial", help="Partial number without check digit")

    args = parser.parse_args()

    if args.cmd == "validate":
        number = args.number.replace(" ", "").replace("-", "")
        if not number.isdigit():
            sys.exit("Error: Input must contain digits only.")
        if not (13 <= len(number) <= 19):
            sys.exit(f"Error: Expected 13–19 digits, got {len(number)}.")
        validate_verbose(number)

    elif args.cmd == "checkdigit":
        partial = args.partial.replace(" ", "").replace("-", "")
        if not partial.isdigit():
            sys.exit("Error: Input must contain digits only.")
        if not (12 <= len(partial) <= 18):
            sys.exit(f"Error: Expected 12–18 digits, got {len(partial)}.")
        checkdigit_verbose(partial)

    else:
        _interactive()


if __name__ == "__main__":
    main()

File: luhn.cpp
/**
 * luhn.cpp — Luhn Algorithm: Checksum Validation & Check Digit Generation
 *
 * Compile:  g++ -std=c++17 -o luhn luhn.cpp
 * Run:      ./luhn
 */

#include <iostream>
#include <string>
#include <vector>
#include <numeric>
#include <algorithm>
#include <cctype>

// -----------------------------------------------------------------------
// Helpers
// -----------------------------------------------------------------------

static bool isDigitsOnly(const std::string& s) {
    return !s.empty() && std::all_of(s.begin(), s.end(), ::isdigit);
}

static std::string sanitize(std::string s) {
    s.erase(std::remove_if(s.begin(), s.end(),
                           [](char c){ return c == ' ' || c == '-'; }), s.end());
    return s;
}

static std::string identifyNumberType(const std::string& num) {
    const size_t len = num.size();
    const std::string p2 = (len >= 2) ? num.substr(0, 2) : "";
    const std::string p4 = (len >= 4) ? num.substr(0, 4) : "";
    const std::string p6 = (len >= 6) ? num.substr(0, 6) : "";

    if (num[0] == '4' && (len == 13 || len == 16 || len == 19))          return "Visa";
    if (num[0] == '5' && len == 16 && p2 >= "51" && p2 <= "55")          return "Mastercard";
    if (num[0] == '2' && len == 16 && p6 >= "222100" && p6 <= "272099")  return "Mastercard (2-series)";
    if (num[0] == '3' && len == 15 && (p2 == "34" || p2 == "37"))        return "American Express";
    if (num[0] == '3' && len == 14 && (p2 == "30" || p2 == "36" || p2 == "38")) return "Diners Club";
    if (len == 16 && (p4 == "6011" || (p2 >= "64" && p2 <= "65")))       return "Discover";
    if (len == 16 && p2 == "62")                                           return "UnionPay";
    if (len == 15 || len == 16)                                            return "IMEI / Credit Card (unrecognized network)";
    return "Unknown";
}

// -----------------------------------------------------------------------
// Mode 1: Validate a complete number
// -----------------------------------------------------------------------

static void runValidate() {
    std::string input;
    std::cout << "Enter the number to validate: ";
    std::cin >> input;
    input = sanitize(input);

    if (!isDigitsOnly(input)) { std::cerr << "\nError: Digits only.\n"; return; }
    if (input.size() < 13 || input.size() > 19) {
        std::cerr << "\nError: Expected 13–19 digits, got " << input.size() << ".\n"; return;
    }

    std::cout << "\nNumber type : " << identifyNumberType(input) << "\n";
    std::cout << "Digit count : " << input.size() << "\n";
    std::cout << "Digits      : ";
    for (char c : input) std::cout << c << " ";

    std::vector<int> d;
    for (char c : input) d.push_back(c - '0');

    // Double every second digit from the right, skipping the check digit
    for (int i = static_cast<int>(d.size()) - 2; i >= 0; i -= 2) d[i] *= 2;
    std::cout << "\n\n[Step 1] Double every second digit from the right:\n         ";
    for (int x : d) std::cout << x << " ";

    for (int& x : d) if (x > 9) x -= 9;
    std::cout << "\n\n[Step 2] Subtract 9 from values > 9:\n         ";
    for (int x : d) std::cout << x << " ";

    const int total = std::accumulate(d.begin(), d.end(), 0);
    std::cout << "\n\n[Step 3] Total sum : " << total << "  |  mod 10 : " << (total % 10)
              << "\n\nResult --> Luhn Checksum: " << (total % 10 == 0 ? "VALID" : "INVALID") << "\n";
}

// -----------------------------------------------------------------------
// Mode 2: Generate check digit for a partial number
// -----------------------------------------------------------------------

static void runCheckDigit() {
    std::string input;
    std::cout << "Enter the partial number (without check digit): ";
    std::cin >> input;
    input = sanitize(input);

    if (!isDigitsOnly(input)) { std::cerr << "\nError: Digits only.\n"; return; }
    if (input.size() < 12 || input.size() > 18) {
        std::cerr << "\nError: Expected 12–18 digits, got " << input.size() << ".\n"; return;
    }

    std::cout << "\nPartial number : ";
    for (char c : input) std::cout << c << " ";
    std::cout << "\nDigit count    : " << input.size() << "\n";

    std::vector<int> d;
    for (char c : input) d.push_back(c - '0');

    // Double starting from the rightmost digit of the partial number
    for (int i = static_cast<int>(d.size()) - 1; i >= 0; i -= 2) d[i] *= 2;
    std::cout << "\n[Step 1] Double every second digit from the right:\n         ";
    for (int x : d) std::cout << x << " ";

    for (int& x : d) if (x > 9) x -= 9;
    std::cout << "\n\n[Step 2] Subtract 9 from values > 9:\n         ";
    for (int x : d) std::cout << x << " ";

    const int total      = std::accumulate(d.begin(), d.end(), 0);
    const int checkDigit = (10 - (total % 10)) % 10;
    std::cout << "\n\n[Step 3] Sum : " << total
              << "  |  Check digit = (10 - " << (total % 10) << ") mod 10 = " << checkDigit
              << "\n\nResult --> Check Digit : " << checkDigit
              << "\n           Full number : " << input << checkDigit << "\n";
}

// -----------------------------------------------------------------------
// main
// -----------------------------------------------------------------------

int main() {
    std::cout << "================================\n";
    std::cout << "  Luhn Algorithm\n";
    std::cout << "================================\n";
    std::cout << "\n[1] Validate a complete number\n";
    std::cout << "[2] Generate a check digit\n";
    std::cout << "\nChoice (1/2): ";

    char choice;
    std::cin >> choice;
    std::cout << "\n";

    switch (choice) {
        case '1': runValidate();   break;
        case '2': runCheckDigit(); break;
        default:  std::cerr << "Invalid choice.\n"; return 1;
    }

    return 0;
}
